### ISSUE TITLE
Parsing strains in GNPS FBMN output needs to be more flexible

### ISSUE THREAD

**Issue by andrewramsay** [NPLinker/nplinker#50]
The [current implementation of parsing FBMN strain labels](https://github.com/sdrogers/nplinker/blob/e094000584f49800aecf852d864503319c7c3f71/prototype/nplinker/metabolomics.py#L411) has turned out be too specific to the Carnegie dataset that we used during development. 

I'd assumed that strainID_growthmedium (e.g. KRD168_ISP3) was a standard naming scheme, but it isn't. The existing parsing code has problems with other labelling methods, and breaks completely if more than 2 underscores are used. 

So the parsing needs to be more flexible both in terms of the structure of the labels and the presence/absence of any growth media information.

**Comment by andrewramsay** [comment]
I've been working on this and have a better solution for it now. But from looking at the GNPS FBMN output it seems like in those cases the information might be available to allow NPLinker to automate a lot of the process of providing strain mappings on the metabolomics side. 

With the Carnegie data, the "metadata_table" file contains one row per strain which gives the full strain identifier (e.g. `KRD77_ISP3.mzML`) and also the growth medium as a separate column (e.g. `ISP3`). 

This isn't really enough on its own because you would likely want to refer to that strain as KRD77. If you search for and remove the ISP3 part it might still leave other arbitrary characters (in this case it would give you `KRD77_`).

However this dataset was using an older version of the GNPS FBMN workflow. It says version "release_14", while the newer dataset I have is version "release_28.2". One difference I've found is that the metadata_table file now has a couple of extra columns, and one of them gives what i guess is the base/fundamental strain name. e.g. for filenames `4_P_aby_KRD168.mzML, 5_P_aby_KRD168.mzML, 6_P_aby_KRD168.mzML`, the "strain" column is always just `KRD168`. 

So if this file is available, NPLinker could do an initial scan of it, take each line, chop off the .mzML part of the filename, create a Strain object from the value in each "Strain" column and add an alias for every filename that matches it. For the above example this would result in a single NPLinker "Strain" object with label `KRD168`, but it would also have aliases called `4_P_aby_KRD168`, `5_P_aby_KRD168`, etc. 

These could be merged into any existing strain_mappings file and potentially save a lot of time entering them manually, and it would also remove the need to guess at parsing the growth medium as the current version of NPLinker does. 

This seems like a good approach but I wanted to check if it makes sense before I implemented it, any thoughts? Are the columns in this file are likely to always be available when using the FBMN workflow? I had a quick look at the source code for the workflow and it seemed like the metadata table might be an input file, so if it is where does it originally come from? Might be useful to know which tool is responsible for formatting it!

**Comment by justinjjvanderhooft** [comment]
Hi @andrewramsay, I think it is a good plan to make such a parser. However, the input for the file names is in the end the machine operator, and it is often the case for the metadata file as well (or the PhD student / PostDoc working on the project) and both names and metadata info is highly "human-dependable".... (i.e., some use hyphens, others not, etc. - in some cases the file name is just a random number and there is another column in the acquisition table that details the sample information) - Thus, a general parser that checks for "the basics", i.e., mzML or mzXML extension, and then looks if there is a metadata file and parses the info from there, and then matches the available info to the strain correlation file is our "safest bet". If we describe the preferred file naming / metadata info needed, we can of course encourage folks to follow that as it will save them time.... Makes sense?

**Comment by andrewramsay** [comment]
OK, sounds like I was being a bit too optimistic about the content of that file being useful in all cases!

Should I actually make the parsing a bit less "smart" and rely on the user to do more? What it's supposed to do right now is take something like `KRD77_ISP3.mzML` and extract a strain ID of `KRD77` (which should appear in strain_mappings.csv) and a growth medium of `ISP3`. 

But since that's not going to work with arbitrary labels, should NPLinker even be trying to parse out the growth media at all if there's not a reliable way to do it? Should it just assume you'll have an entry for `KRD77_ISP3` in strain_mappings.csv and treat that as a strain label without any further parsing?

**Comment by justinjjvanderhooft** [comment]
I guess it will only work with some descriptions from our side as to what our code expects. Downside for the user: take care with the naming (or rename files prior to analysis); upside: more automation within NPLinker.... 

In all other cases I would just use a "sensibility filter" to check for file formats etc. and then rely on the strain correlation file to be as flexible as possible....

**Comment by andrewramsay** [comment]
What I've done for now is have a default where the parsing is extremely simple, it just takes the filename minus the .mz(X)ML part and assumes that is the strain name that should be in the strain mappings file. 

But I've also added a new config file option described on [this new wiki page](https://github.com/sdrogers/nplinker/wiki/Strain-mappings#strain-naming-schemes-fbmn-workflows) which if enabled (it defaults to being disabled) will try to parse the other columns in the metadata table file as I suggested above. For the dataset I have here this seems to work nicely but if it defaults to being disabled then it shouldn't behave unexpectedly on other datasets. 

I hope this is a reasonable solution for the moment!
